### PR TITLE
Stick breaking process [WIP]

### PR DESCRIPTION
**Pull request by Harivallabha**
[WIP] Creating pull request so that Austin and Max can review the structure of the implementation.
(and the logp method)

random and _random methods haven't been implemented correctly yet [WIP].
Have to write tests.

### CODE REVIEW

File: pymc3/dp/util.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    def __init__(self, a, num_comp, transform=transforms.stick_breaking,
+                 *args, **kwargs):
+        shape = num_comp   # Num_comp should be equal to the number of weights
+        kwargs.setdefault("shape", shape)
+        super().__init__(transform=transform, *args, **kwargs)
+        self.a = tt.as_tensor_variable(a)
+        self.k = tt.as_tensor_variable(shape)
+        self.mean = 1.  # Just to get stuff running. Testing the logp right now.
+
+    def _random(self, size=None):
```
Comment (ferrine) [inline pymc3/dp/util.py:51]: This is not working, need to implement it correctly and check with a test

File: pymc3/dp/util.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        wts = draw_values([self.a], point=point, size=size)
+        samples = generate_samples(self._random,
+                                   wts=wts,
+                                   dist_shape=self.shape,
+                                   size=size)
+        return samples
+
+    def get_betas(self, weights):
+        betas, _ = theano.scan(fn=lambda prior_beta, index, weights: (prior_beta * weights[index] / (1 - prior_beta) * weights[index - 1]),
```
Comment (ferrine) [inline pymc3/dp/util.py:81]: `scan` will work slow, better to do some math to obtain a closed form vectorized solution.

**Review by ferrine — CHANGES_REQUESTED** [review]

File: pymc3/dp/util.py (hunk 2 above)
Comment (AustinRochford) [reply pymc3/dp/util.py:81]: The [Stan manual](https://mc-stan.org/docs/2_18/reference-manual/simplex-transform-section.html) gives an explicit formula for the inverse of the stickbreaking transform (see the bottom of that link) that can be implemented using just `cumsum` (apologies for suggesting `cumprod`)

**Comment by twiecki** [comment]
Closing due to inactivity, feel free to reopen.

**Pull request closed without merging**